### PR TITLE
Add se-edu branding + navigation bar to website

### LINKED ISSUE #736: Add se-edu branding + navigation bar to website
### Problem

Visitors to our [website](https://https://se-edu.github.io/addressbook-level4/), when browsing one of the subpages of the website, e.g. [UsingGradle](https://se-edu.github.io/addressbook-level4/UsingGradle.html), cannot immediately tell that they are browsing the addressbook-level4 website.

Visitors also have no way of quickly jumping to frequently used pages, e.g. the User Guide, Developer Guide, or homepage.

### Suggested Solution

Add SE-EDU Addressbook-level4 branding to the top of the page, as well as a navigation bar with links to frequently-visited pages like the homepage, user guide, and developer guide.

### Requirements

* Navigation bar links: I think ideally we'll need 2 levels, one level for the addressbook-levelX projects, and another level for the addressbook-level4 specific links. Until the navigation bar is implemented in all addressbook websites though, I think we can just stick to one.
    * Links should follow the site map: User Guide, Developer Guide, Learning Outcomes, About Us, Contact Us.
    * (Nice to have) The corresponding link should be highlighted if the user is viewing that page. e.g., in [User Guide](https://se-edu.github.io/addressbook-level4/UserGuide.html), the "User Guide" link should be highlighted. This highlighting should work with subpages too, e.g. when viewing [UsingGradle](https://se-edu.github.io/addressbook-level4/UsingGradle.html) the "Developer Guide" link should be highlighted.
* The navigation bar should not show up in the print view, since we use the print view to generate PDFs.
* The branding, and navigation bar links, should ideally be easily modifiable, so that downstream projects can modify them.
* The navigation bar links should work on both the hosted github pages website and when viewing the locally generated version. i.e., use relative URLs, not absolute ones.
* I don't think the branding and navigation bar will need to be displayed when the User Guide is shown in the help window in the application, otherwise it will allow the user to travel to different pages and break the immersion of the help window :-P
* The implemented solution should, as much as possible, not require developers to do any additional installation steps.
* Try to avoid adding boilerplate to all asciidoc files in the implemented solution.
* The branding and navigation bar should look decent enough :-)

### PR DESCRIPTION
**Pull request by pyokagan**
Fixes #736 

```
We would like to give visitors easy access to the AddressBook-Level4
documentation as well as other SE-EDU projects, in order to build a
cohesive SE-EDU brand image.

Add two navigation bars to the top of the AddressBook-Level4 website --
one SE-EDU navigation bar that links to other SE-EDU projects, and one
site navigation bar that provides easy access to AddressBook-Level4's
User Guide, Developer Guide etc.
```

### CODE REVIEW

**Comment by pyokagan** [comment]
Argh, forgot: this includes some code from the asciidoctor project, and so the `LICENSE` file needs to be updated to include a copy of its license in accordance with MIT License rules.

**Comment by damithc** [comment]
@pyokagan Sorry I didn't think of this before: perhaps we should present AB4 site as an independent product site? i.e., not have the first row in the nav bar. That's because students will use this to create their own product site. The top row is more appropriate for the se-edu site rather than AB4 site, don't you think?

**Comment by damithc** [comment]
> @damithc There is a configuration option to remove the SE-EDU bar (see [9/12]).

Should be fine then. 👍

File: docs/templates/_header.html.slim [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+              =nav_link('LearningOutcomes', 'LearningOutcomes.html', 'LOs')
+          li.nav-item
+            =nav_link('AboutUs', 'AboutUs.html', 'About Us')
+          li.nav-item
+            =nav_link('ContactUs', 'ContactUs.html', 'Contact Us')
+          - if attr? 'site-githuburl'
+            li.navitem
+              a.nav-link href=(attr 'site-githuburl')
+                span.fa.fa-github.fa-lg aria-hidden='true'
+                | &nbsp;GitHub
```
Comment (pyokagan) [inline docs/templates/_header.html.slim]: Currently this looks like it's a link to github.com.

I think the link text should be `View on GitHub`.

**Comment by pyokagan** [comment]
Ah, figured out why Travis didn't want to trigger, I had a PR open to my own personal fork, so Travis probably got confused -- https://github.com/pyokagan/addressbook-level4/pull/4

**Comment by pyokagan** [comment]
Changes:

* Some Developer Guide tweaks
* Made the navigation bar look nicer in small screen sizes -- the navigation bar title (i.e. the SE-EDU logo or `AddressBook-Level4`) will now be centered.
* `View on GitHub`
* Added LICENSE file for imported asciidoctor-backend code

**Comment by pyokagan** [comment]
@yamgent @yamidark btw, I'd like to merge this by Friday if possible, since I'll be overseas after that.

I understand you all are busy, but I'm not demanding a full review either -- just glance through the commit messages to check that the intent behind makes sense, read through the developer guide and maybe play around with it? I'm quite confident that the code will work.

(Unlike se-edu-bot which died due to a lack of reviews (and I'm fine with that), this PR is a bit too important to die :-P )

File: docs/DeveloperGuide.adoc [hunk 2]
```diff
@@ -420,6 +427,67 @@ Here are the steps to convert the project documentation files to PDF format.
...
+|Attribute name |Description |Default value
+
+|`site-name`
+|The name of the website. This will be displayed at the top of the page.
+|`AddressBook-Level4`
+
+|`site-githuburl`
+|URL to the site's repository on https://github.com[GitHub].
+Setting this will add a "View on GitHub" link in the navigation bar.
+|_not set_
```
Comment (yamgent) [inline docs/DeveloperGuide.adoc]: Default is set to `https://github.com/se-edu/addressbook-level4`.

File: docs/DeveloperGuide.adoc [hunk 3]
```diff
@@ -420,6 +427,67 @@ Here are the steps to convert the project documentation files to PDF format.
...
+
+|`site-seedu`
+|Define this attribute if the project is an official SE-EDU project.
+This will render the SE-EDU navigation bar at the top of the page, and add some SE-EDU-specific navigation items.
+|_not set_
```
Comment (yamgent) [inline docs/DeveloperGuide.adoc]: Default is set to `true`.

File: docs/DeveloperGuide.adoc [hunk 4]
```diff
@@ -68,7 +69,13 @@ Optionally, you can follow the <<UsingCheckstyle#, UsingCheckstyle.adoc>> docume
 
 ==== Updating documentation to match your fork
 
-After forking the repo, links in the documentation will still point to the `se-edu/addressbook-level4` repo. If you plan to develop this as a separate product (i.e. instead of contributing to the `se-edu/addressbook-level4`) , you should replace the URL in the variable `repoURL` in `DeveloperGuide.adoc` and `UserGuide.adoc` with the URL of your fork.
+After forking the repo, the documentation will still refer to the `se-edu/addressbook-level4` repo.
```
Comment (yamgent) [inline docs/DeveloperGuide.adoc]: Can also mention to the students about the additional site navigation bar that is specific to SE-EDU.

If necessary, the `site-seedu` attribute can also be mentioned here to make the connection between the nav-bar and the attribute more obvious.

**Comment by pyokagan** [comment]
@yamgent 
> Add instructions on how to restart the gradle daemon. Rationale: Students may not know how to restart the gradle daemon, and may spend ages debugging this issue.

So, here's the thing. I strongly *discourage* students from modifying the template files, especially if they are not experienced enough with programming. If students are not able to figure out what the gradle damon is and how to restart it such that they "spend ages debugging the issue", it's also an indication that they should not be modifying the template files, which requires some knowledge of Ruby and asciidoctor APIs, and even so, are *extremely easy* to break. (hence the warning on not modifying the template files unless CSS stylesheets are not powerful enough)

Well, maybe I shouldn't be passing premature judgement, so here's what I will do:

```diff
diff --git a/docs/DeveloperGuide.adoc b/docs/DeveloperGuide.adoc
index 940fd8f..d46b6de 100644
--- a/docs/DeveloperGuide.adoc
+++ b/docs/DeveloperGuide.adoc
@@ -488,6 +488,11 @@ The files in link:{repoURL}/docs/templates[`docs/templates`] controls the render
 These template files are written in a combination of https://www.ruby-lang.org[Ruby] and http://slim-lang.com[Slim].
 You should only modify them if you need greater control over the site's layout than what stylesheets can provide.
 
+[NOTE]
+====
+The Gradle daemon must be <<UsingGradle#RestartDaemon, restarted>> after modifying any of the files in link:{repoURL}/docs/templates[`docs/templates`] for the changes to take effect.
+====
+
 [[Testing]]
 == Testing
 
diff --git a/docs/UsingGradle.adoc b/docs/UsingGradle.adoc
index d1be2f3..6b68061 100644
--- a/docs/UsingGradle.adoc
+++ b/docs/UsingGradle.adoc
@@ -111,3 +111,14 @@ See `build.gradle` ->
 * **`compileTestJava`** +
 Checks whether the project has the required dependencies to perform testing, and download any missing dependencies before compiling the test classes. +
 See `build.gradle` -> `allprojects` -> `dependencies` -> `testCompile` for the list of dependencies required.
+
+[[RestartDaemon]]
+== Restarting the Gradle daemon
+
+By default, Gradle will spawn a long-lived background process (daemon) to avoid the cost of JVM startup for every build.
+This background process will also cache information about project structure, files, tasks, etc. in memory.
+
+If you'd like to invalidate the cache, you can restart the Gradle daemon as follows:
+
+. Stop the Gradle daemon by running `./gradlew --stop`.
+. Start the Gradle daemon again by running any task.
```

**Comment by pyokagan** [comment]
> So, here's the thing. I strongly discourage students from modifying the template files, 

What I actually don't want is students modifying the template files, seeing them break, and then blaming us for it (filing an issue and such, which has happened before).

So, I think I'll add this:

```diff
diff --git a/docs/DeveloperGuide.adoc b/docs/DeveloperGuide.adoc
index d46b6de..c353a6b 100644
--- a/docs/DeveloperGuide.adoc
+++ b/docs/DeveloperGuide.adoc
@@ -486,7 +486,13 @@ You can modify them to change some styling properties of the site.
 
 The files in link:{repoURL}/docs/templates[`docs/templates`] controls the rendering of `.adoc` files into HTML5.
 These template files are written in a combination of https://www.ruby-lang.org[Ruby] and http://slim-lang.com[Slim].
+
+[WARNING]
+====
+Modifying the template files in link:{repoURL}/docs/templates[`docs/templates`] requires some knowledge and experience with Ruby and Asciidoctor's APIs.
 You should only modify them if you need greater control over the site's layout than what stylesheets can provide.
+The SE-EDU team provides no support for any modified template files.
+====
 
 [NOTE]
 ====

```

I'll be dropping the stuff about restarting the gradle daemon, since otherwise it'll be unnecessary extra documentation to maintain.

File: docs/DeveloperGuide.adoc (hunk 2 above)
Comment (pyokagan) [reply docs/DeveloperGuide.adoc]: Not really. Will this help clarify the documentation?

```diff
diff --git a/docs/DeveloperGuide.adoc b/docs/DeveloperGuide.adoc
index 32c4319..028af2c 100644
--- a/docs/DeveloperGuide.adoc
+++ b/docs/DeveloperGuide.adoc
@@ -443,7 +443,10 @@ The `build.gradle` file specifies some project-specific https://asciidoctor.org/
 
 |`site-githuburl`
 |URL to the site's repository on https://github.com[GitHub].
+
 Setting this will add a "View on GitHub" link in the navigation bar.
+
+Leaving this unset will not add a "View on GitHub" link in the navigation bar.
 |_not set_
 
 |`site-seedu`
```

Comment (pyokagan) [reply docs/DeveloperGuide.adoc]: Oh nevernmind, I think I see the problem now.

How about something like this?

```diff
diff --git a/docs/DeveloperGuide.adoc b/docs/DeveloperGuide.adoc
index 32c4319..8c3e42a 100644
--- a/docs/DeveloperGuide.adoc
+++ b/docs/DeveloperGuide.adoc
@@ -432,6 +432,9 @@ image::chrome_save_as_pdf.png[width="300"]
 
 The `build.gradle` file specifies some project-specific https://asciidoctor.org/docs/user-manual/#attributes[asciidoc attributes] which affects how all documentation files within this project are rendered.
 
+[TIP]
+Attributes left unset in the `build.gradle` file will use their *default value*.
+
 [cols="1,2a,1", options="header"]
 .List of site-wide attributes
 |===
```

File: docs/DeveloperGuide.adoc (hunk 4 above)
Comment (pyokagan) [reply docs/DeveloperGuide.adoc]: The comment in `build.gradle` should be sufficient, main thing is just to direct devs to `build.gradle`, and everything else will be self-evident there.

I'll make the following changes:

```diff
diff --git a/docs/DeveloperGuide.adoc b/docs/DeveloperGuide.adoc
index 04156d1..7734925 100644
--- a/docs/DeveloperGuide.adoc
+++ b/docs/DeveloperGuide.adoc
@@ -69,7 +69,7 @@ Optionally, you can follow the <<UsingCheckstyle#, UsingCheckstyle.adoc>> docume
 
 ==== Updating documentation to match your fork
 
-After forking the repo, the documentation will still refer to the `se-edu/addressbook-level4` repo.
+After forking the repo, the documentation will still refer to SE-EDU and the `se-edu/addressbook-level4` repo.
 
 If you plan to develop this fork as a separate product (i.e. instead of contributing to `se-edu/addressbook-level4`), you should do the following:
 
```

**Comment by pyokagan** [comment]
@yamidark 
> Commit message: extra whitespace in docs/templates/ directory?

Sorry, I don't quite get this. Did you mean it should be `docs/templates/directory`? `docs/templates/` is correct.

> Commit title is more than 72 chars. Perhaps can just be docs/templates: remind devs when to restart the gradle daemon?

Done

**Comment by yamidark** [comment]
>Sorry, I don't quite get this. Did you mean it should be docs/templates/directory? docs/templates/ is correct.

Whoop sorry, my mistake.

**Comment by pyokagan** [comment]
@yamgent @yamidark Thanks for the reviews. I really appreciate it.

**Pull request merged**